external-reports/src/main/java/org/sonar/java/externalreport/pmd/PmdXmlReportReader.java

```diff
@@ -0,0 +1,161 @@
+      String elementName = element.getName().getLocalPart();
+      if ("file".equals(elementName)) {
+        String filePath = getAttributeValue(element, "name");
+        FilePredicates predicates = context.fileSystem().predicates();
+        inputFile = context.fileSystem().inputFile(predicates.hasPath(filePath));
+        if (inputFile == null) {
+          LOG.warn("No input file found for {}. No PMD issue will be imported on this file.", filePath);
+        }
+      } else if ("violation".equals(elementName) && inputFile != null) {
+        onViolationStartElement(element);
+      }
+
+    } else if (event.isCharacters()) {
+      issueMessage.append(event.asCharacters().getData());
+
+    } else if (event.isEndElement()
+      && "violation".equals(event.asEndElement().getName().getLocalPart())
+      && inputFile != null
+      && issue != null) {
+      issueLocation.message(issueMessage.toString());
+      issue.at(issueLocation).save();
+    }
+  }
+
+  private void onViolationStartElement(StartElement element) {
+    try {
+      TextRange textRange = textRange(element);
+      RuleKey ruleKey = RuleKey.of(PmdSensor.LINTER_KEY, getAttributeValue(element, "rule"));
+      issue = context.newExternalIssue()
+        .forRule(ruleKey)
+        .type(RuleType.CODE_SMELL)
+        .severity(SEVERITIES.get(getAttributeAsInt(element, "priority")))
+        .remediationEffortMinutes(ruleLoader.ruleConstantDebtMinutes(ruleKey.rule()));
+      issueLocation = issue.newLocation()
+        .on(inputFile)
+        .at(textRange);
+      issueMessage = new StringBuilder();
+    } catch (RuntimeException e) {
+      int lineNumber = element.getLocation().getLineNumber();
+      LOG.warn("Can't import issue at line " + lineNumber + " in " + reportFile + ": " + e.getMessage());
+      issue = null;
+    }
+  }
+
+  private TextRange textRange(StartElement violationElement) {
+    Integer beginLine = getAttributeAsInt(violationElement, "beginline");
+    try {
+      Integer endLine = getAttributeAsInt(violationElement, "endline");
+      Integer beginColumn = getAttributeAsInt(violationElement, "begincolumn");
+      Integer endColumn = getAttributeAsInt(violationElement, "endcolumn");
+      return inputFile.newRange(beginLine, beginColumn - 1, endLine, endColumn);
+    } catch (RuntimeException e) {
+      // Some PMD rules seem to report invalid line offsets, e.g. TooManyStaticImports
+      return inputFile.selectLine(beginLine);
+    }
+  }
+
+  private static String getAttributeValue(StartElement startElement, String attributeName) {
+    Attribute attribute = startElement.getAttributeByName(new QName(attributeName));
+    return attribute == null ? "" : attribute.getValue();
+  }
+
+  private static Integer getAttributeAsInt(StartElement startElement, String attributeName) {
+    return Integer.parseInt(getAttributeValue(startElement, attributeName));
+  }
+
+  private static Map<Integer, Severity> severities() {
```

Comment:
isn't it better to put this code into static initializer block

Reply:
Why would it be better?

Reply:
then it's explicit that this method is intended to be used only for initialization

Reply:
It's not intended to be used only for initialization.